### PR TITLE
Support Bot new invoke type: config

### LINKED ISSUE #6634: Support New Bot Invoke Type: Config
Support new bot invoke type: config, including config/submit and config/fetch.

The bot config response for config/submit or config/fetch could be one of the following:
- config/auth
- config/continue
- config/message
- Terminate with 200.

### PR DESCRIPTION
**Pull request by yingduyingdu**
Fixes #6634

## Description
Support new bot invoke type: config.

## Specific Changes
- Support new config invoke requests, including config/submit and config/fetch.
- Support related bot schema for Bot config responses.

## Testing
E2E manual tests against local changes for various config requests and responses.
Unit Tests.

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll**

File: libraries/Microsoft.Bot.Builder/Teams/TeamsActivityHandler.cs [hunk 1]
```diff
@@ -431,6 +437,32 @@ protected virtual Task<TabResponse> OnTeamsTabSubmitAsync(ITurnContext<IInvokeAc
...
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A Config Response for the request.</returns>
+        protected virtual Task<InvokeResponseBase> OnTeamsConfigFetchAsync(ITurnContext<IInvokeActivity> turnContext, JObject configData, CancellationToken cancellationToken)
+        {
+            throw new InvokeResponseException(HttpStatusCode.NotImplemented);
+        }
+
+        /// <summary>
+        /// Override this in a derived class to provide logic for when a config is fetched.
```
Comment (FBX622) [inline libraries/Microsoft.Bot.Builder/Teams/TeamsActivityHandler.cs:454]: minor: should be config is submitted.

File: libraries/Microsoft.Bot.Builder/Teams/TeamsActivityHandler.cs [hunk 2]
```diff
@@ -431,6 +437,32 @@ protected virtual Task<TabResponse> OnTeamsTabSubmitAsync(ITurnContext<IInvokeAc
...
+        /// </summary>
+        /// <param name="turnContext">A strongly-typed context object for this turn.</param>
+        /// <param name="configData">The config fetch invoke request value payload.</param>
```
Comment (FBX622) [inline libraries/Microsoft.Bot.Builder/Teams/TeamsActivityHandler.cs:457]: "submit"

File: libraries/Microsoft.Bot.Schema/Teams/BotConfigAuth.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Schema.Teams
+{
+    /// <summary>
+    /// Specifies bot config authe, including type and suggestedActions.
```
Comment (FBX622) [inline libraries/Microsoft.Bot.Schema/Teams/BotConfigAuth.cs:8]: typo: "auth"

File: libraries/Microsoft.Bot.Schema/Teams/InvokeResponseBase.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+// Licensed under the MIT License.
+
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Schema.Teams
+{
+    /// <summary>
+    /// Specifies Invoke response base including response type.
+    /// </summary>
+    public partial class InvokeResponseBase
```
Comment (FBX622) [inline libraries/Microsoft.Bot.Schema/Teams/InvokeResponseBase.cs:11]: Is there a reason to name it as InvokeResponseBase, will we have a class of InvokeResponse? I saw we have both classes for task.

**Review by corinagum — APPROVED** [review]

File: libraries/Microsoft.Bot.Builder/Teams/TeamsActivityHandler.cs (hunk 1 above)
Comment (yingduyingdu) [reply libraries/Microsoft.Bot.Builder/Teams/TeamsActivityHandler.cs:454]: Updated.

File: libraries/Microsoft.Bot.Schema/Teams/BotConfigAuth.cs (hunk 3 above)
Comment (yingduyingdu) [reply libraries/Microsoft.Bot.Schema/Teams/BotConfigAuth.cs:8]: Will update in the next iteration

File: libraries/Microsoft.Bot.Schema/Teams/InvokeResponseBase.cs (hunk 4 above)
Comment (yingduyingdu) [reply libraries/Microsoft.Bot.Schema/Teams/InvokeResponseBase.cs:11]: This will align with our class design in the reference repo.

Comment (yingduyingdu) [reply libraries/Microsoft.Bot.Schema/Teams/InvokeResponseBase.cs:11]: I can sync up with you offline for this.

File: tests/Microsoft.Bot.Builder.Tests/Teams/TeamsActivityHandlerTests.cs [hunk 5]
```diff
@@ -1052,6 +1052,72 @@ void CaptureSend(Activity[] arg)
...
+            }
+
+            var turnContext = new TurnContext(new SimpleAdapter(CaptureSend), activity);
+
+            // Act
+            var bot = new TestActivityHandler();
+            await ((IBot)bot).OnTurnAsync(turnContext);
+
+            // Assert
+            //Assert.Equal(2, bot.Record.Count);
```
Comment (FBX622) [inline tests/Microsoft.Bot.Builder.Tests/Teams/TeamsActivityHandlerTests.cs:1112]: this is commented.

File: tests/Microsoft.Bot.Builder.Tests/Teams/TeamsActivityHandlerTests.cs [hunk 6]
```diff
@@ -1052,6 +1052,72 @@ void CaptureSend(Activity[] arg)
...
```
Comment (FBX622) [inline tests/Microsoft.Bot.Builder.Tests/Teams/TeamsActivityHandlerTests.cs:1079]: this is commented out.

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll**

File: tests/Microsoft.Bot.Builder.Tests/Teams/TeamsActivityHandlerTests.cs (hunk 5 above)
Comment (yingduyingdu) [reply tests/Microsoft.Bot.Builder.Tests/Teams/TeamsActivityHandlerTests.cs:1112]: Updated in the next iteration

File: tests/Microsoft.Bot.Builder.Tests/Teams/TeamsActivityHandlerTests.cs (hunk 6 above)
Comment (yingduyingdu) [reply tests/Microsoft.Bot.Builder.Tests/Teams/TeamsActivityHandlerTests.cs:1079]: Updated in the next iteration

**Review by FBX622 — APPROVED** [review]

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll**

**Comment by yingduyingdu** [comment]
The PR will NOT be merged before the Bot config front-end changes rollout to R3. 
It will take a couple of days to weeks to roll out to R3 from now.

**Review by ceciliaavila — APPROVED** [review]
LGTM

**Review by tracyboehrer — APPROVED** [review]

**Pull request merged**